### ISSUE TITLE
"out-of-memory"-error in initializer with high dimensional search space

### ISSUE THREAD

**Issue by SimonBlanke** [SimonBlanke/Gradient-Free-Optimizers#1]
In issue SimonBlanke/Hyperactive#8 a problem was found, where the memory usage can increase rapidly.

The error is located in the "_init_search"-method from the "Search"-class.

On systems with ~12gb of ram an "out-of-memory"-error will occur when searching in a high dimensional (>20) search space.

**Comment by SimonBlanke** [comment]
I managed to identify the reason for the error. In the file "init_positions.py" there are two functions with this problem.

They are called:
 - _init_grid_search
 - _init_vertices

They initialize positions at the vertices of the search space hypercube and in a grid.

The reason the memory error occurs is because "np.meshgrid(...)" creates too many possible positions for a large number of dimensions. All possible vertex positions are calculated and from this collection GFO then selects n of those positions (where n is initialize={"vertices": n} from the api).
It would be much better if "_init_vertices" adds vertex positions one by one into a list and returns them in the function.

"_init_grid_search" has a similar problem. It will probably be more difficult to reimplement, because I don't like how those positions are selected at the moment + there is the memory-error.

**Comment by SimonBlanke** [comment]
Hello @architrathore,

Yes it would be good to start with `_init_vertices`. Your approach seems good! Remember the following when writing this function:

Those points are just positions in the search spaces (integers), not the values from the search space dictionary. So this dictionary:
```python
search_space = {
    "x1": np.arange(-10, 10, 1),
    "x2": np.arange(-10, 10, 1),
}
```
would have the following list of vertex positions:
```python
[array([19, 19]), array([19,  0]), array([ 0, 19]), array([0, 0])]
```

And try to stick to the following template for the function:
```python
# n_pos is "n" from initialize={"vertices": n}

def _init_vertices(self, n_pos):
    positions = your_algorithm(...)

    # if n is higher than the maximum number of vertices, fill the rest with random positions
    positions = self._fill_rest_random(n_pos, positions)

    return positions
```

**Comment by romanovzky** [comment]
This is a major issue for me. Why can't spaces be defined either through a callable or a distribution (like scipy distributions with rvs methods)? I've found this to be common practice in every other gradient-free and Bayesian optimisation libraries (hyperopt, optuna, scikit-opt, etc).

**Comment by SimonBlanke** [comment]
Hello @romanovzky,

could you provide a short code snippet as an example for one (or more) of those libraries?

**Comment by romanovzky** [comment]
Hi @SimonBlanke sure, here are some examples on how these libraries support this.

## SciPy

SciPY defines the most common API for having the distribution as a function from which you can sample at will. This funcitonality is provided by the `rvs` (random variable sample) method, [example](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.rv_continuous.rvs.html).

[All the distributions in SciPy](https://docs.scipy.org/doc/scipy/reference/stats.html) follow this logic. For example a [loguniform](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.loguniform.html#scipy.stats.loguniform) and [uniform](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.uniform.html#scipy.stats.uniform).

This helps encoding the whole scan space without making a grid (which you do directly with the `np.arange`).

## Optuna

A [`trial` object](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.trial.Trial.html#optuna.trial.Trial) is passed as the argument of the objective function. The `trial` then can `suggest` values according to the distribution type and its bounds. The sampler then suggests from the available spaces that the trial presents to it. [Here](https://github.com/optuna/optuna/blob/9fe44349c05c9da38e4c8e84e9e048037e701862/optuna/samplers/_tpe/multivariate_parzen_estimator.py#L81) you can find the TPE sample method.

You see that it makes use of `rvs` methods, with the parameters of the distributions being learnt by the TPE.

## HyperOpt

Their codebase is a mess, but they usel `pyll` to define the spaces as distributions instead of arrays of examples, [here](https://github.com/hyperopt/hyperopt/blob/master/hyperopt/pyll/stochastic.py).

Their approach is similar to yours: they define the space before calling the objective function (as opposed to optuna). You can see that the workflow [here](https://github.com/hyperopt/hyperopt/blob/master/tutorial/02.MultipleParameterTutorial.ipynb) is very similar to your own.

## Scikit-optimize

They follow a similar approach again, where the space can be defined beforehand using distribution assignments. See [their Space API](https://scikit-optimize.github.io/stable/modules/classes.html#module-skopt.space.space). Again, notice how a dimension (i.e. a variable) has a `rvs` method.

For initial samplers, I'd recommend going through the `scikit-optimize` [`samplers` API](https://scikit-optimize.github.io/stable/modules/classes.html#module-skopt.sampler), where you can see different approaches to independent/warm-up samplings.

## Other examples

- [SMAC](https://automl.github.io/SMAC3/master/quickstart.html) also uses pre-defined variable bounds, within which it samples/operates.
- [Hyperas](https://github.com/maxpumperla/hyperas) as well, although it's not surprising as it is built in top of hyperopt.

## Final comments

All the cases above are Bayesian optimisation libraries, so these define the prior space over the variable space and the TPE, Gaussian Process, etc learn over these priors. But since they use the full prior (as oppose to a grid over it, as your methodology using `np.arange` imposes) you don't need to form these super huge grids.

The libraries I shown above, which I know better, are very focused on ML hyper-parameter tuning. I now have a different problem and would need grad-free optimisers (this is how I came across your library) and I am testing a few (I'm adapting my optuna, hyperopt, etc workflows to fit this new task) but I was very keen on trying the wealth of optimisers from your work.

**Comment by SimonBlanke** [comment]
Thank you very much @romanovzky for taking your time writing down this overview! From what I read it seams, that your question is about the API-design and continuous- vs- discrete- search spaces. I think it would be better to open up a new issue to discuss this.

**Comment by SimonBlanke** [comment]
Hello @architrathore,

did you manage to implement the new version of [_init_vertices](https://github.com/SimonBlanke/Gradient-Free-Optimizers/blob/master/gradient_free_optimizers/init_positions.py#L91)? If you are stuck I would be happy to help. 

The reimplementation of _init_grid_search can wait, since it is ambiguous how to implement it. I will take care of this function.

**Comment by architrathore** [comment]
Hi @SimonBlanke, apologies but I got side-tracked by other work, but I did have an implementation that worked. I will create a pull request in the next week.

**Comment by SimonBlanke** [comment]
@architrathore,
I am glad you were able to take some time for this project. It can be difficult to free up time for open source projects. However I look forward to your pull request.

**Comment by SimonBlanke** [comment]
I created a fix for the init-vertices part of the problem. For the init-grid I built in a workaround for dimensions higher than 30 in commit 70604fefd9c61ac79ecd01d793660b32f9d86267. Those fixes are available in v1.0.5.

I will close this issue for now. I think a grid-initialization for more than 30 dimensions is a niche case and should not make a significant difference. If this becomes relevant in the future I can reopen this issue.
